prov/util/src/util_buf.c

```diff
@@ -45,6 +45,44 @@ enum {
 };
 
 
+static int ofi_bufpool_alloc(struct ofi_bufpool_region *buf_region)
+{
+	int ret;
+	struct ofi_bufpool *pool = buf_region->pool;
+
+	if (pool->attr.flags & OFI_BUFPOOL_HUGEPAGES) {
+		ret = ofi_alloc_hugepage_buf((void **) &buf_region->alloc_region,
+					     pool->alloc_size);
+		if (!ret) {
+			buf_region->flags = OFI_BUFPOOL_HUGEPAGES;
+			return 0;
+		}
+		/* If we can't allocate huge pages, fall back to normal
+		 * allocations for all future attempts.
+		 */
+		pool->attr.flags &= ~OFI_BUFPOOL_HUGEPAGES;
+	}
+
+	if (pool->attr.flags & OFI_BUFPOOL_MMAP) {
```

Comment:
Other locations use else if after checking for huge pages, which is what I think you want here as well.  hugepages and mmap are exclusive from each other.

Reply:
This is on purpose.  I originally wrote this to fall through to mmap if huge page allocation failed, but decided to leave this to follow-up discussion.

The only time efa asks for huge pages is when fork support is not needed, so in theory it would be safe to just use small pages as before.  However, it may be nice to move the buffers involved with `ibv_reg_mr` into a separate mmap anyway.  This could be accomplished by keeping the fall-through here and specifying `OFI_BUFPOOL_HUGEPAGES | OFI_BUFPOOL_MMAP` in efa (see https://github.com/nzmsv/libfabric/pull/1).

Thoughts?

Reply:
Hmm... huge page allocation is via mmap.  So I'm thinking that the definition should be:
```
OFI_BUFPOOL_MMAP = 1 << 3,
OFI_BUFPOOL_HUGEPAGES = (1 << 4) | OFI_BUFPOOL_MMAP,
```
The fallback from huge pages mmap -> malloc could be considered an error since we're no longer guaranteed to return page aligned allocations.  (Even falling back from mmap to malloc is questionable in that regard, but so unlikely to happen that I think we're okay.)

Reply:
Basically, I think your original choice to fallback to mmap is what we want.

Reply:
I decided to keep the MMAP and HUGEPAGE flags separate because we may want to unset HUGEPAGE individually (for example, when it fails, or when the allocation is smaller than the size of a single hugepage) but keep the mmap behavior.